### ISSUE TITLE
Signature of fromShaun

### ISSUE THREAD

**Issue by Docteur-Lalla** [Shumush/SNIHs#1]
Hello,

the signature of the `fromShaun` function in Shaun class states that a convertion from a `ShaunValue` to the corresponding instance always success. This is obviously false.

Maybe the function should have type `fromShaun :: ShaunValue -> Maybe a` or a similar functor or the class should have another function that can fail instead of returning a default value.

For instance, when deserializing a complex object, one has to return a default (obviously erroneous) value in case of failure:

```haskell
import Data.List as L

data Name = Name { firstname :: String, lastname :: String }

instance Shaun Name where
  toShaun name = SObject
    [ ("firstname", toShaun $ firstname name)
    , ("lastname", toShaun $ lastname name)
    ]

  fromShaun (SObject attrs) = case nameOpt of
    Just name -> name
    Nothing -> -- what now ?
    where
      nameOpt = do
        first <- L.lookup "firstname" attrs
        last <- L.lookup "lastname" attrs
        return $ Name
          { firstname = fromShaun first
          , lastname = fromShaun last
          }

  fromShaun _ = -- what now ?
```

**Comment by lilymonad** [comment]
You are right, and while I am at it, why not generalizing to `fromShaun :: (MonadThrow m) => ShaunValue -> m a`. As `Maybe` and `Either e` instantiate this class, we can even use a special `ShaunException` to specify what is going on.

**Comment by ablanleuil** [comment]
Commit #295ed0a6ddeda2995345659a02f71119bb5c4a2d fixed this issue.

**Comment by lilymonad** [comment]
Fixed in commit 295ed0a6ddeda2995345659a02f71119bb5c4a2d.
The function `fromShaun` is now `(MonadThrow m) => ShaunValue -> m a`.

You now have to use `throwM` to throw a `ShaunException`, now defined in `Types.hs`:
```Haskell
data ShaunException
  = OutOfRange
  | AttributeNotFound
  | NotAList
  | NotAnObject
  | NotAString
  | NotANumber
  | NotABool
  | Custom String
  deriving (Show)
```
